src/toolchain_override.rs

```diff
@@ -0,0 +1,218 @@
+                    None
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn get_component_version(&self, component: &str) -> Option<&Version> {
+        if let Some(components) = &self.cfg.components {
+            components.get(component)
+        } else {
+            None
+        }
+    }
+
+    pub fn install_missing_components(&self, toolchain: &Toolchain, called: &str) -> Result<()> {
+        match &self.cfg.components {
+            None => warn!(
+                "warning: overriding toolchain '{}' in {} does not have any components listed",
+                &self.cfg.toolchain.channel, FUEL_TOOLCHAIN_TOML_FILE
+            ),
+            Some(components) => {
+                for component in components.keys() {
+                    if !toolchain.has_component(component) {
+                        let target_triple = TargetTriple::from_component(component)?;
+
+                        if let Ok(download_cfg) = DownloadCfg::new(called, target_triple, None) {
+                            info!(
+                                "installing missing component '{}' specified in {}",
+                                component, FUEL_TOOLCHAIN_TOML_FILE
+                            );
+                            toolchain.add_component(download_cfg)?;
+                        };
+                    }
+                }
+            }
+        };
+        Ok(())
+    }
+}
+
+impl OverrideCfg {
+    pub fn new(toolchain: ToolchainCfg, components: Option<HashMap<String, Version>>) -> Self {
+        Self {
+            toolchain,
+            components,
+        }
+    }
+
+    // Creates a representation of a 'fuel-toolchain.toml' from a toml string.
+    // This is used in the implementation of ToolchainOverride, which is just
+    // an OverrideCfg with its file path.
+    pub(crate) fn from_toml(toml: &str) -> Result<Self> {
+        let cfg: OverrideCfg = de::from_str(toml)?;
+        if DistToolchainName::from_str(&cfg.toolchain.channel).is_err() {
+            bail!("Invalid channel '{}'", &cfg.toolchain.channel)
+        }
+
+        if let Some(components) = cfg.components.as_ref() {
+            if components.is_empty() {
+                bail!("'[components]' table is declared with no components")
+            }
+        }
+
+        Ok(cfg)
+    }
+
+    pub fn to_document(self) -> Document {
+        let mut document = toml_edit::Document::new();
+
+        document["toolchain"] = toml_edit::Item::Table(toml_edit::Table::new());
+        document["toolchain"]["channel"] = value(self.toolchain.channel);
+        if let Some(components) = self.components {
+            document["components"] = toml_edit::Item::Table(toml_edit::Table::new());
+            for (k, v) in components.iter() {
+                document["components"][k] = value(v.to_string());
+            }
+        }
+
+        document
+    }
+}
```

Comment:
Rather than having manual `from_toml` and `to_document` implementations, is it possible to have the derived `Deserialize` or `Serialize` do the conversions as we'd expect?

Reply:
The derived `Deserialize`/`Serialize` outputs them not in the format that we want. 

I dug a little into the `toml_edit` docs and seems like `to_string_pretty()` ([link](https://docs.rs/toml_edit/latest/toml_edit/ser/fn.to_string_pretty.html)) supports the use case that I want for tests here.